### PR TITLE
make api rate limit test a little more robust

### PR DESCRIPTION
**Pull request by DaanHoogland**
### Description

This PR...

Fixes: #6975 

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [x] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [x] Major
- [ ] Minor
- [ ] Trivial

### How Has This Been Tested?

unit test

### CODE REVIEW

**Review by nvazquez — APPROVED** [review]

**Review by davidjumani — APPROVED** [review]
CLGTM

**Comment by davidjumani** [comment]
@blueorangutan package

File: plugins/api/rate-limit/src/test/java/org/apache/cloudstack/ratelimit/ApiRateLimitTest.java [hunk 1]
```diff
@@ -110,17 +110,25 @@ public void sequentialApiAccess() {
...
-        assertFalse("We should block >" + allowedRequests + " requests per second", isUnderLimit(key));
+        // we cannot really say more about this test
+        boolean underLimit =  isUnderLimit(key);
+        long endtime = System.nanoTime();
+        int issued = s_limitService.getIssued(key.getAccountId());
+        int timeToLive = s_limitService.getTimeToLive();
+
+        // this assertion is really invalid as we don´t know if we exceeded the time to live for the amount of api calls (for sure)
+        assertFalse(String.format("We should block >%d requests per %d seconds (managed %d, time elapsed %d)",
+                s_limitService.getMaxAllowed(), timeToLive, issued, endtime - startTime), underLimit);
```
Comment (DaanHoogland) [inline plugins/api/rate-limit/src/test/java/org/apache/cloudstack/ratelimit/ApiRateLimitTest.java:131]:
```suggestion
        // this assertion is really invalid as we don´t know if we exceeded the time to live for the amount of api calls (for sure)
        // so only fail if timeToLive is not exeeded and we didn´t get the requested number of calls
        assertFalse(String.format("We should block >%d requests per %d seconds (managed %d, time elapsed %d)",
                s_limitService.getMaxAllowed(), timeToLive, issued, endtime - startTime), underLimit || ((endtime - starttime)/1000000 > timeToLive);
```

Comment (DaanHoogland) [reply plugins/api/rate-limit/src/test/java/org/apache/cloudstack/ratelimit/ApiRateLimitTest.java:131]: @davidjumani @nvazquez @weizhouapache , I did a self review and I think we need this extra condition

**Comment by DaanHoogland** [comment]
@blueorangutan package

File: plugins/api/rate-limit/src/test/java/org/apache/cloudstack/ratelimit/ApiRateLimitTest.java [hunk 2]
```diff
@@ -110,17 +110,26 @@ public void sequentialApiAccess() {
 
     @Test
     public void canDoReasonableNumberOfApiAccessPerSecond() throws Exception {
-        int allowedRequests = 200;
+        int allowedRequests = 2000;
```
Comment (DaanHoogland) [inline plugins/api/rate-limit/src/test/java/org/apache/cloudstack/ratelimit/ApiRateLimitTest.java:113]:
```suggestion
        int allowedRequests = 200;
```

File: plugins/api/rate-limit/src/test/java/org/apache/cloudstack/ratelimit/ApiRateLimitTest.java [hunk 3]
```diff
@@ -110,17 +110,26 @@ public void sequentialApiAccess() {
...
+        // we cannot really say more about this test
+        boolean underLimit =  isUnderLimit(key);
+        long endtime = System.nanoTime();
+        int issued = s_limitService.getIssued(key.getAccountId());
+        int timeToLive = s_limitService.getTimeToLive();
+
+        // this assertion is really invalid as we don´t know if we exceeded the time to live for the amount of api calls (for sure)
+        // so only fail if timeToLive is not exeeded and we didn´t get the requested number of calls
+        assertTrue(String.format("We should block >%d requests per %d seconds (managed %d, time elapsed %d)",
+                s_limitService.getMaxAllowed(), timeToLive, issued, endtime - startTime), ( ! underLimit) || ((endtime - starttime)/1000000 > timeToLive);
```
Comment (DaanHoogland) [inline plugins/api/rate-limit/src/test/java/org/apache/cloudstack/ratelimit/ApiRateLimitTest.java:132]:
```suggestion
        // this assertion is really invalid as we don´t know if we exceeded the time to live for the amount of api calls (for sure)
        // so only fail if timeToLive is not exeeded and we didn´t get the requested number of calls
        assertFalse(String.format("We should block >%d requests per %d seconds (managed %d, time elapsed %d)",
                s_limitService.getMaxAllowed(), timeToLive, issued, endtime - startTime), ((endtime - starttime)/1000000 < timeToLive) && underLimit);
```

**Review by DaanHoogland — COMMENTED** [review]
some cleenup

File: plugins/api/rate-limit/src/test/java/org/apache/cloudstack/ratelimit/ApiRateLimitTest.java [hunk 4]
```diff
@@ -112,15 +112,24 @@ public void sequentialApiAccess() {
...
 
         for (int i = 0; i < allowedRequests; i++) {
-            assertTrue("We should allow " + allowedRequests + " requests per second, but failed at request " + i, isUnderLimit(key));
+            assertTrue(String.format("We should allow %d requests per second, but failed at request %d.", allowedRequests, i), isUnderLimit(key));
         }
-
-        assertFalse("We should block >" + allowedRequests + " requests per second", isUnderLimit(key));
+        // we cannot really say more about this test
+        boolean underLimit =  isUnderLimit(key);
+        long endtime = System.nanoTime();
```
Comment (weizhouapache) [inline plugins/api/rate-limit/src/test/java/org/apache/cloudstack/ratelimit/ApiRateLimitTest.java:125]:
```suggestion
        long endtime = System.nanoTime();
        System.out.println("time elapsed " + (endtime - startTime)/1000/1000 + " ms");
```

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
test succeed after 12 seconds, looks good.

```
20:39:01 [INFO] Running org.apache.cloudstack.ratelimit.ApiRateLimitTest
20:39:06 SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
20:39:06 SLF4J: Defaulting to no-operation (NOP) logger implementation
20:39:06 SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
20:39:06 log4j:WARN No appenders could be found for logger (org.apache.cloudstack.ratelimit.ApiRateLimitServiceImpl).
20:39:06 log4j:WARN Please initialize the log4j system properly.
20:39:06 log4j:WARN See http://logging.apache.org/log4j/1.2/faq.html#noconfig for more info.
20:39:10 time elapsed 3347 ms
20:39:13 [INFO] Tests run: 7, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 12.544 s - in org.apache.cloudstack.ratelimit.ApiRateLimitTest
```

**Review by weizhouapache — APPROVED** [review]

**Pull request merged**